### PR TITLE
Introduce new approach for UI testing framework

### PR DESCRIPTION
**Pull request by OleksandrOrlov**
**For reviewers: Please, merge needles first before merging the PR!**

**Motivation:**
Currently we have an implementation of UI test framework that:
- Do not allow to easily identify what test module actually does, especially if the test module is large;
- Has a lot of conditional statements in test modules to separate between versions of products;
- Hardly allows to reuse existing steps in test scenarios with another expected result;
- Has a lot of flacky tests.

**Goal:**
Design a prototype of a new UI testing framework that will solve all the above and apply it for an existing test module (it was decided to choose 'partitioning_lvm' test module).
(Please see [poo#44753](https://progress.opensuse.org/issues/44753) to have brief understanding of what planned to be achieved).

- Related ticket: https://progress.opensuse.org/issues/44753
- **Needles:** 
opensuse: https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/510
sle: https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/1067

- Verification runs: 
  - SLE15 lvm encrypt: http://10.160.65.138/tests/821#step/lvm_encrypt/9
  - SLE15 no lvm encrypt: http://10.160.65.138/tests/820
  - Tumbleweed lvm: http://10.160.65.138/tests/825
  - SLE12 lvm encrypt: http://10.160.65.138/tests/819
  - SLE12 lvm: http://10.160.65.138/tests/787

### CODE REVIEW

File: lib/PageInteractions.pm [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+use Exporter;
+
+our @EXPORT = qw(
+    find
+    send_keys
+    fill_in_password
+    is_displayed
+);
+
+sub find {
```
Comment (rwx788) [inline lib/PageInteractions.pm]: It's more like `wait`. I'm not sure about this particular thing.

File: lib/PageInteractions.pm [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    my $self = @_;
```
Comment (rwx788) [inline lib/PageInteractions.pm]: We don't need this line I guess.

File: lib/PageInteractions.pm [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    my $key = shift;
+    my $message = shift || "Send $key key";
+    record_info('send_key', $message);
+    send_key($key, @_);
+}
+
+sub fill_in_password {
+    my $password = shift;
+    my $message = shift || "Type password";
+    my $self = @_;
```
Comment (rwx788) [inline lib/PageInteractions.pm]: We follow `my ($self) = @_;` style everywhere else, don't you mind to change it? We don't need `$self` here either.

File: lib/PageInteractions.pm [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    record_info('type_password', $message);
+    type_password($password, @_);
+}
+
+sub is_displayed {
+    my $needle = shift;
+    my $timeout = shift || 10;
+    record_info('check_screen', "Check whether $needle area exists on visible screen");
```
Comment (rwx788) [inline lib/PageInteractions.pm]: I guess this particular wrapper doesn't help us, as usual `check_screen` provides same level of the information.

File: lib/opensusebasetest.pm [hunk 5]
```diff
@@ -651,4 +651,10 @@ sub post_fail_hook {
     save_screenshot;
 }
 
+sub assert_true {
```
Comment (rwx788) [inline lib/opensusebasetest.pm]: I wonder if we can actually reuse what is already in `Test` and `Test::More` packages.

File: lib/DistributionProvider.pm [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+use version_utils;
+
+use Distribution::Tumbleweed;
+use Distribution::Sle15Sp1;
+use Distribution::Sle12Sp4;
+
+sub provide {
+    return Distribution::Tumbleweed->new() if version_utils::is_tumbleweed();
+    return Distribution::Sle15Sp1->new() if version_utils::is_sle('=15-SP1');
+    return Distribution::Sle12Sp4->new() if version_utils::is_sle('=12-SP4');
```
Comment (rwx788) [inline lib/DistributionProvider.pm]: I guess we can still define ranges, as not we'll get undef for Leap and SP5, which ATM is same as SP4.

File: lib/Installation/Partitioner/Libstorage/ProposalSettingsDialog.pm [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+);
+
+my $PROPOSAL_SETTINGS_DIALOG = 'inst-partition-radio-buttons';
+
+sub select_encrypted_lvm_based_proposal_radiobutton {
+    find($PROPOSAL_SETTINGS_DIALOG);
+    send_keys('alt-e', 'Select "Encrypted LVM-based Proposal" radiobutton.');
+}
+
+sub apply_proposal_settings {
```
Comment (jknphy) [inline lib/Installation/Partitioner/Libstorage/ProposalSettingsDialog.pm]: Perhaps call it press_ok or something like that to avoid confuse with "apply" buttons.

File: lib/Installation/Partitioner/LibstorageNG/TooSimplePasswordDialog.pm [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+use PageInteractions;
+
+use parent 'Exporter';
+use Exporter;
+
+our @EXPORT = qw(
+    agree_with_too_simple_password
+);
+
+my $TOO_SIMPLE_PASSWORD_DIALOG = 'inst-userpasswdtoosimple';
```
Comment (jknphy) [inline lib/Installation/Partitioner/LibstorageNG/TooSimplePasswordDialog.pm]: I guess lower case at least you aim to introduce some constant.

File: lib/PageInteractions.pm [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+sub find {
+    my $self = @_;
+    assert_screen(@_);
+}
+
+sub send_keys {
+    my $key = shift;
+    my $message = shift || "Send $key key";
+    record_info('send_key', $message);
```
Comment (jknphy) [inline lib/PageInteractions.pm]: I would write the shortcut instead of the name of the function. It is quite visually annoying to read many of all those boxes with "send_keys", but $message is really useful. Other option, just use "Press key" but I guess it would kind of overkill me visually. Perhaps we need some enhancement in os-autoinst with this boxes, instead of creating a box, in the point when the api is used to be able to add the description in $message.

File: tests/installation/partitioning_encrypted_lvm.pm [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+use testapi;
+use parent "installbasetest";
+
+sub run {
+    my $self = shift;
+    my $partitioner = $testapi::distri->get_partitioner();
+
+    $partitioner->encrypt_partition_with_too_simple_password(is_lvm => 1);
+
+    $self->assert_true($partitioner->is_encrypted_partition_with_lvm_shown_in_the_list(),
```
Comment (jknphy) [inline tests/installation/partitioning_encrypted_lvm.pm]: OOP should allows to create shorter names. Perhaps it is for the level of specialization but it is too long in my opinion.

File: tests/installation/partitioning_encrypted_no_lvm.pm [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+use strict;
+use warnings FATAL => 'all';
+use testapi;
+use parent "installbasetest";
+
+sub run {
+    my $self = shift;
+    my $partitioner = $testapi::distri->get_partitioner();
+
+    $partitioner->encrypt_partition_with_too_simple_password(is_lvm => 0);
```
Comment (jknphy) [inline tests/installation/partitioning_encrypted_no_lvm.pm]: Also too long name for the function, I guess more OOP oriented would be to call it `encrypt_partition` and pass password and another param for example `complexity -> simple` or use a number and some comments for possible values.

**Comment by jknphy** [comment]
Looks really promising, good job @OleksandrOrlov . 
I really like the part of the providers to separate logic among products. Perhaps the OOP for each particular test could done in a way to avoid those long name in the methods, but thinking twice, that part is the most open one in the sense that anyone could decide what is the level of specialization to use when implementing classes and also developers could still write code in the current way if I get it right.

File: lib/PageInteractions.pm (hunk 1 above)
Comment (OleksandrOrlov) [reply lib/PageInteractions.pm]: My initial idea was to make something like find('needle_of_the_screen_where_action_should_be_performed')->send_key('smth'), mainly to sort of _force_ test writers to carefully check where the test is currently running before performing an action (that causes a lot of troubles currently).

My initial plan was to make testapi_extended class, that would be a child of testapi  with the 'find' method, that returns something like FoundArea class, where all the actions could be done. But current os-autoinst implementation don't allow to do this, as testapi is a perl package, not the class. Also, it is not really clear what to pass to FoundArea class as os-autoinst also not designed that way. So I just added this find() method as-is to show my vision on how the framework might be extended.

But for now, I agree that name for the method might be different (_ensure_visible_ for example). Again, I just want to ask test developers to check a screen before performing an action.

File: lib/PageInteractions.pm (hunk 2 above)
Comment (OleksandrOrlov) [reply lib/PageInteractions.pm]: I wanted to delegate all the method parameters to assert_screen to keep same flexibility as assert_screen has. Not sure if I did this correctly in terms of Perl.

File: lib/PageInteractions.pm (hunk 3 above)
Comment (OleksandrOrlov) [reply lib/PageInteractions.pm]: Again my poor Perl skills, I wanted to pass the same arguments to type_password method and add some specific parameters (as $message) to the extended _fill_in_password_ and didn't find quick solution how to do this. Probably @_[1]

File: lib/PageInteractions.pm (hunk 4 above)
Comment (OleksandrOrlov) [reply lib/PageInteractions.pm]: Again, as with other subs in this package, my poor Perl skills. Wanted to pass same arguments to record_info, but also add $message to 'is_displayed'. Did not achieve that, so left some default message, but I would like to improve that in case we'll proceed with the solution.

File: lib/opensusebasetest.pm (hunk 5 above)
Comment (OleksandrOrlov) [reply lib/opensusebasetest.pm]: Yes, good point. Just didn't know about that method, but wanted to show in the prototype what I want to achieve. 

Will try to use this.

File: lib/DistributionProvider.pm (hunk 6 above)
Comment (OleksandrOrlov) [reply lib/DistributionProvider.pm]: For the prototype it was enough to introduce TW and both Sle versions. 
For working solution, just one line with Leap distri will do the trick.return Distribution::Leap15_1->new() if version_utils::is_leap('=15.1');

File: lib/Installation/Partitioner/LibstorageNG/TooSimplePasswordDialog.pm (hunk 8 above)
Comment (OleksandrOrlov) [reply lib/Installation/Partitioner/LibstorageNG/TooSimplePasswordDialog.pm]: Yes, wanted to introduce constant. 

The main idea was to keep needles in the top of a package as constants, so that it's easy to see how many needles are used in the particular module and to allow easily manage them in one place in case it they need to be changed.

File: lib/DistributionProvider.pm (hunk 6 above)
Comment (rwx788) [reply lib/DistributionProvider.pm]: I would simply put `return Distribution::Tumbleweed->new();` in the end. Also, soften conditions to `15-SP1+` and `12-SP4+`

Comment (rwx788) [reply lib/DistributionProvider.pm]: As for naming convention, I would use something like Distribution::SLE::12::SP4 Distribution::SLE::15::SP1 or without extra layer Distribution::SLE::12SP4, but I guess there will be common part between all service packs, so for SP0 we'll use  Distribution::SLE::12. WDYT? We also could reuse VERSION variable to find appropriate perl module.

File: lib/Installation/Partitioner/LibstorageNG/TooSimplePasswordDialog.pm (hunk 8 above)
Comment (jknphy) [reply lib/Installation/Partitioner/LibstorageNG/TooSimplePasswordDialog.pm]: right, I think for `my` variables is lower letter and for constants we can group them like this:
```
use constant {
    FOO => "bar",
    BAR => "foo"
};
```

File: lib/PageInteractions.pm (hunk 3 above)
Comment (okurz) [reply lib/PageInteractions.pm]: @OleksandrOrlov Take a look in https://github.com/os-autoinst/os-autoinst/tree/master/testapi.pm . I am sure you will find good examples for parameter parsing including the according documentation and "support for legacy behaviour" in there.

File: lib/main_common.pm [hunk 12]
```diff
@@ -865,10 +865,10 @@ sub load_inst_tests {
                 loadtest "installation/partitioning_raid";
             }
             elsif (check_var('LVM', 0) && get_var('ENCRYPT')) {
-                loadtest 'installation/partitioning_crypt_no_lvm';
+                loadtest 'installation/partitioning_encrypted_no_lvm';
             }
             elsif (get_var("LVM")) {
-                loadtest "installation/partitioning_lvm";
+                loadtest "installation/partitioning_encrypted_lvm";
```
Comment (okurz) [inline lib/main_common.pm]: I think you are changing the meaning here in an unintended way. This one here is probably covering lvm in both cases, unencrypted *and* encrypted

File: lib/PageInteractions.pm [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    fill_in_password
+    is_displayed
+);
+
+sub find {
+    my $self = @_;
+    assert_screen(@_);
+}
+
+sub send_keys {
```
Comment (dzedro) [inline lib/PageInteractions.pm]: Can't testapi send_key have this feature instead of complication ?

File: lib/PageInteractions.pm (hunk 1 above)
Comment (dzedro) [inline lib/PageInteractions.pm]: Creating new function which will do the same as old function is nonsense
I like that you want to make some thing more obvious, but this "wrappers" make openQA just more complicated why don't you add this feature directly into send_key, type_password, check_screen and so on ?

**Comment by sysrich** [comment]
@okurz my biggest fear with this different approach is the practical problems it could introduce in debugging what went wrong with a test and why

Currently the openQA webui shows you the source code in the webui for each test module. This is intended to give you a good reference of what the test module was doing so it should be trivial to figure out "Oh, that's what it was trying to do" so you can figure out what went wrong with the failure

We _already_ have a problem with all of the code abstracted in various "_common.pm" modules, because these are not discoverable in the WebUI, so from a regular reviewer you are left with significant gaps about the what openQA was intended to do missing. The logs in this case are not good enough, you need a source describing the tests _intent_ as well as the logs which provide you with a source of what actually happened.

With this approach, I fear the above existing problem would be exacerbated to the point where it would be almost impossible to ever figure out what the test what was actually doing from the WebUI as it currently stands.

So, my suggestion/request before heavily considering adoption of this approach would be to work on the WebUI so it easily provides the reviewer a clearer picture of what functions the test was trying to do, and which *.pm that function was called from, at every step of the test flow

This I think would eliminate my concerns to the point where this highly abstracted approach to test writing wouldn't get in my way too much

I don't think I'd want to adopt it for the Kubic tests though..but we do typically have simpler tests than some of the more complex ones out there :)

**Comment by okurz** [comment]
@sysrich has a point, thanks for sharing.

File: lib/PageInteractions.pm (hunk 1 above)
Comment (OleksandrOrlov) [reply lib/PageInteractions.pm]: Thanks for the comment. Yes, you're right, but I just wanted to show what I would like to have in openQA and added just dump wrappers to get the feedback about the whole idea of having every step somehow logged on openQA job overview page.

I totally agree that it should be implemented in os-autoinst and openQA should also allow to show the steps in more user-friendly and clear way, rather then using 'record_info' squares for each step.

File: lib/main_common.pm [hunk 14]
```diff
@@ -2426,4 +2426,23 @@ sub load_transactional_role_tests {
     loadtest 'transactional_system/rebootmgr';
 }
 
+sub load_lvm_tests {
```
Comment (rwx788) [inline lib/main_common.pm]: Could you, please, add couple of comments and ideally describe variables.

File: products/opensuse/main.pm [hunk 15]
```diff
@@ -69,9 +70,7 @@ sub cleanup_needles {
 my @time = localtime();
 set_var('WINTER_IS_THERE', 1) if ($time[4] == 11 || $time[4] == 0);
 
-my $distri = testapi::get_required_var('CASEDIR') . '/lib/susedistribution.pm';
-require $distri;
-testapi::set_distribution(susedistribution->new());
+testapi::set_distribution(DistributionProvider->provide());
```
Comment (rwx788) [inline products/opensuse/main.pm]: I believe we still need to set distri to `susedistribution` for old cases.

Comment (rwx788) [reply products/opensuse/main.pm]: Ok, actually it's not an issue, as used in all.

File: lib/Distribution/Opensuse/Leap.pm [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+# Follows the "Factory first" rule. So that the feature first appears in
+# Tumbleweed distribution, and only if it behaves different in Leap then it
+# should be overriden here.
+
+# Maintainer: Oleksandr Orlov <oorlov@suse.de>
+
+package Distribution::Opensuse::Leap;
+use strict;
+use warnings FATAL => 'all';
+use parent 'Distribution::Opensuse::Tumbleweed';
```
Comment (okurz) [inline lib/Distribution/Opensuse/Leap.pm]: Actually Leap is derived from SLE *and* Tumbleweed depending on the component so encoding it as simple as that is not true. For the installer however you can expect it to be similar to SLE regarding functionality but us the openSUSE theming which should only be visible in needles.

File: lib/main_common.pm [hunk 17]
```diff
@@ -859,11 +859,11 @@ sub load_inst_tests {
...
             }
-            elsif (check_var('LVM', 0) && get_var('ENCRYPT')) {
-                loadtest 'installation/partitioning_crypt_no_lvm';
+            elsif (get_var('LVM')) {
+                load_lvm_tests();
             }
-            elsif (get_var("LVM")) {
-                loadtest "installation/partitioning_lvm";
+            elsif (check_var('LVM', 0) && get_var('ENCRYPT')) {
+                loadtest 'installation/partitioning/no_lvm_encrypt';
```
Comment (okurz) [inline lib/main_common.pm]: I suggest to keep the old name and leave that for a separate PR. IMHO "no_lvm_encrypt" can be misleading as meaning "no_lvm" but "encrypt" or "no_lvm" and also "no_encrypt"

File: lib/main_common.pm [hunk 18]
```diff
@@ -2426,4 +2426,29 @@ sub load_transactional_role_tests {
...
+# reconfigures lvm partition, otherwise partition suggestion based on existing
+# cryptlvm should compute encrypted system (currently it refers to bsc#993247).
+sub load_lvm_tests {
+    if (get_var("ENCRYPT")) {
+        if (get_var('ENCRYPT_ACTIVATE_EXISTING')) {
+            if (get_var('ENCRYPT_FORCE_RECOMPUTE')) {
+                loadtest 'installation/partitioning/lvm_encrypt_recompute_existing';
+            }
+            else {
+                loadtest 'installation/partitioning/lvm_encrypt_should_compute_existing';
```
Comment (okurz) [inline lib/main_common.pm]: Are you sure this long module name is still displayed nicely?

File: lib/main_common.pm (hunk 14 above)
Comment (OleksandrOrlov) [reply lib/main_common.pm]: Added.

File: lib/main_common.pm (hunk 17 above)
Comment (OleksandrOrlov) [reply lib/main_common.pm]: Namings changed to be more clear.

File: lib/main_common.pm (hunk 18 above)
Comment (OleksandrOrlov) [reply lib/main_common.pm]: Namings changed to be more clear.

File: lib/Distribution/Opensuse/Leap.pm (hunk 16 above)
Comment (OleksandrOrlov) [reply lib/Distribution/Opensuse/Leap.pm]: So, here I would go with the existing solution as for now. As for the partitioning lvm it completes all the needs. In case if sometime we'll need to extend hierarchy, then it could be done. 

One smart guy just recently explained me "Never Add Functionality Early" principle;)

**Comment by rwx788** [comment]
> The test modules read very nicely now however I feel we might make some people unhappy with a change as such when the source code view from openQA is not handled in a nicer way. As already suggested in person: Why not work on that feature on the openQA side first before going further here? We can still manage this as a team :)

I would emphasize that we try to make it step by step and fail fast. With "fix the world" we will postpone real change and increase investment in the idea we are not 100% sure about. So I will push further, we try it out, revert and reconsider if it doesn't work. We already drag this relatively simple change for a month now. All of us want to see things moving instead of finding yet another reason not to move forward. And I can assure you, there always will be more than one.

Richard has mentioned valid point, which exist regardless of this change and we need to address. Obviously, we would not experience any issues if our coverage would be for single distribution. I also can admit that for kubic `main.pm` is not any issue and works nicely, but that can change over time in case we start covering more backend, more versions, etc.

We gave enough time for people to express their opinions. On top this functionality is quite specific and is under QSF control only.
